src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1119,281 @@ private static string RegularExpressionFromFileSpec
+        ///
+        /// (1) The leading \\ in UNC paths, so that the doubled slash isn't reduced in the last step
+        /// 
+        /// (2) Common filespec characters
+        /// </summary>
+        private static void AppendRegularExpressionFromFixedDirectory(ReuseableStringBuilder regex, string fixedDir)
+        {
+            regex.Append(FileSpecRegexParts.BeginningOfLine);
+
+            bool isUncPath = NativeMethodsShared.IsWindows && fixedDir.Length > 1
+                             && fixedDir[0] == '\\' && fixedDir[1] == '\\';
+            if (isUncPath)
             {
-                matchFileExpression.Replace(separator, "<:dirseparator:>");
+                regex.Append(FileSpecRegexParts.UncSlashSlash);
             }
+            int startIndex = isUncPath ? IndexOfNextNonCollapsibleChar(fixedDir, 2) : 0;
 
-            /*
-             * Capture the leading \\ in UNC paths, so that the doubled slash isn't
-             * reduced in a later step.
-             */
-            matchFileExpression.Replace("<:fixeddir:><:dirseparator:><:dirseparator:>", "<:fixeddir:><:uncslashslash:>");
-
-            /*
-             * Iteratively reduce four cases involving directory separators
-             * 
-             *  (1) <:dirseparator:>.<:dirseparator:> -> <:dirseparator:>
-             *        This is an identity, so for example, these two are equivalent,
-             * 
-             *            dir1\.\dir2 == dir1\dir2
-             * 
-             *    (2) <:dirseparator:><:dirseparator:> -> <:dirseparator:>
-             *      Double directory separators are treated as a single directory separator,
-             *      so, for example, this is an identity:
-             * 
-             *          f:\dir1\\dir2 == f:\dir1\dir2
-             * 
-             *      The single exemption is for UNC path names, like this:
-             * 
-             *          \\server\share != \server\share
-             * 
-             *      This case is handled by the <:uncslashslash:> which was substituted in
-             *      a prior step.
-             * 
-             *  (3) <:fixeddir:>.<:dirseparator:>.<:dirseparator:> -> <:fixeddir:>.<:dirseparator:>
-             *      A ".\" at the beginning of a line is equivalent to nothing, so:
-             * 
-             *          .\.\dir1\file.txt == .\dir1\file.txt
-             * 
-             *  (4) <:dirseparator:>.<:eol:> -> <:eol:>
-             *      A "\." at the end of a line is equivalent to nothing, so:
-             * 
-             *          dir1\dir2\. == dir1\dir2             *
-             */
-            int sizeBefore;
-            do
+            for (int i = startIndex; i < fixedDir.Length; i = IndexOfNextNonCollapsibleChar(fixedDir, i + 1))
             {
-                sizeBefore = matchFileExpression.Length;
+                AppendRegularExpressionFromChar(regex, fixedDir[i]);
```

Comment:
This method also checks for characters that should never appear in the fixed directory party, which could make it confusing to understand. How about composing `AppendRegularExpressionFromChar` from 2 methods, one that operators on non-wildcard characters and one that operators on wildcard characters?

Reply:
It turns into 3 methods since slashes also never appear in filenames, so it'd really be a separate method for each part only sharing the ```sb.Append()``` of escaped and normal characters, and at that point I'd just stick it in the parent call. So code duplication vs clarity?

Reply:
Yea, no code duplication :)
One option would be to make it configurable from the caller. Either a bool parameter for each special character, or an array, or an array of structs containing the input character to be matched and it's corresponding regex to be appended.